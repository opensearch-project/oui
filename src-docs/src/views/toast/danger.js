/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import { OuiToast } from '../../../../src/components';

const esError =
  'Error: expected _scroll_id in the following response: {"took":0,"timed_out":false,"_shards":{"total":0,"successful":0,"skipped":0,"failed":0},"hits":{"total":0,"max_score":0,"hits":[]}}';
export default () => (
  <OuiToast
    title="Couldn't complete the search"
    color="danger"
    iconType="alert">
    <p>{esError}</p>
  </OuiToast>
);
