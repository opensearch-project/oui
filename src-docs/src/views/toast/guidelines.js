/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { useState } from 'react';

import { GuideRule, GuideRuleExample, GuideRuleTitle } from '../../components';

import {
  OuiText,
  OuiButton,
  OuiSpacer,
  OuiFlexGroup,
  OuiFlexItem,
  OuiToast,
  OuiCallOut,
  OuiModal,
  OuiModalBody,
  OuiModalFooter,
  OuiModalHeader,
  OuiModalHeaderTitle,
  OuiCodeBlock,
} from '../../../../src/components';

export default () => {
  const [isModalVisible, setIsModalVisible] = useState(false);

  const closeModal = () => {
    setIsModalVisible(false);
  };

  const showModal = () => {
    setIsModalVisible(true);
  };

  let modal;

  if (isModalVisible) {
    modal = (
      <OuiModal onClose={closeModal}>
        <OuiModalHeader>
          <OuiModalHeaderTitle>
            <h1>Your visualization has an error</h1>
          </OuiModalHeaderTitle>
        </OuiModalHeader>

        <OuiModalBody>
          <OuiCallOut
            title="The maximum bucket size of 200 was exceeded"
            color="danger"
            size="s"
            iconType="alert"
          />
          <OuiSpacer size="s" />
          <OuiCodeBlock>
            {`--- FAKE ERROR ---
An extremely long error trace can exist in a modal so you have time
and space to read it properly. Alternatively just link to a full page.
---
                `}
          </OuiCodeBlock>
        </OuiModalBody>

        <OuiModalFooter>
          <OuiButton onClick={closeModal} fill>
            Close
          </OuiButton>
        </OuiModalFooter>
      </OuiModal>
    );
  }

  return (
    <>
      <OuiText grow={false}>
        <p>
          This page documents patterns for using toasts, short messages that
          appears on the lower right corner and time out after a few seconds.
          They are a popular design choice because they don&apos;t need to fit
          in a layout and don&apos;t disrupt the user.
        </p>
      </OuiText>

      <OuiSpacer size="xl" />

      <dl>
        <OuiFlexGroup alignItems="center">
          <OuiFlexItem grow={false} style={{ minWidth: 120 }}>
            <OuiToast
              style={{ width: 300 }}
              title="Your report is complete"
              color="success"
            />
          </OuiFlexItem>

          <OuiFlexItem>
            <OuiText>
              <dt>Success toasts indicate that everything worked out</dt>
              <dd>They are the most-commonly used toasts.</dd>
            </OuiText>
          </OuiFlexItem>
        </OuiFlexGroup>

        <OuiSpacer />

        <OuiFlexGroup alignItems="center">
          <OuiFlexItem grow={false} style={{ minWidth: 120 }}>
            <OuiToast
              style={{ width: 300 }}
              title="Node 726 is having trouble"
              color="warning"
            />
          </OuiFlexItem>

          <OuiFlexItem>
            <OuiText>
              <dt>
                Warning toasts direct user attention to a potential problem
              </dt>
              <dd>
                These toasts work well in monitoring apps when something
                significant requires action.
              </dd>
            </OuiText>
          </OuiFlexItem>
        </OuiFlexGroup>

        <OuiSpacer />

        <OuiFlexGroup alignItems="center">
          <OuiFlexItem grow={false} style={{ minWidth: 120 }}>
            <OuiToast
              style={{ width: 300 }}
              title="Search failed. Check your connection to the cluster."
              color="danger"
            />
          </OuiFlexItem>

          <OuiFlexItem>
            <OuiText>
              <dt>Error toasts report a problem</dt>
              <dd>
                An error toast might let users know an action didn&apos;t
                complete or that a form has errors.
              </dd>
            </OuiText>
          </OuiFlexItem>
        </OuiFlexGroup>

        <OuiSpacer />

        <OuiFlexGroup alignItems="center">
          <OuiFlexItem grow={false} style={{ minWidth: 120 }}>
            <div>
              <OuiToast
                style={{ width: 300 }}
                title="Please wait while your report is created"
                color="primary"
              />
            </div>
          </OuiFlexItem>

          <OuiFlexItem>
            <OuiText>
              <dt>Info toasts relay neutral information</dt>
              <dd>
                The default toast, an info toast might notify users about an
                ongoing action.
              </dd>
            </OuiText>
          </OuiFlexItem>
        </OuiFlexGroup>
      </dl>

      <OuiSpacer />

      <GuideRuleTitle>Use a toast for a timely message</GuideRuleTitle>

      <GuideRule
        description="Toasts are appropriate for short feedback related to a user action.
          A toast should contain a message about a current action, not a historical action.">
        <GuideRuleExample
          type="do"
          text="Use a toast for a brief message about the current action.">
          <OuiToast
            style={{ maxWidth: 300 }}
            title="Your folder was moved"
            color="success"
          />
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          text="Greet users with a toast when they open a page.">
          <div style={{ textAlign: 'center' }}>
            <OuiToast
              style={{ maxWidth: 300 }}
              title="Haven't seen you in a while"
              color="primary"
            />
          </div>
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>
        Most often, it&apos;s a single line of text
      </GuideRuleTitle>

      <GuideRule
        description="By default, a toast stays on the screen 10 seconds.
          Users should be able read the message in 6 to 7 seconds.
          The message should get straight to the point and rarely include more than one line.

          ">
        <GuideRuleExample
          type="do"
          text="A single line of text is readable at a glance.">
          <div style={{ textAlign: 'center' }}>
            <OuiToast
              style={{ maxWidth: 300 }}
              title="Check your form for errors"
              color="danger"
            />
          </div>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          text="Cram a lot of detail into a toast.
            These errors should persist in callouts and validations on the form.
            They don't need to be spelled out in the toast.">
          <div>
            <OuiToast
              style={{ maxWidth: 300 }}
              title="Your form has errors"
              color="danger">
              <OuiText>
                <ul>
                  <li>Username is a required field.</li>
                  <li>Password must be at least 6 characters long.</li>
                  <li>Email is a required field.</li>
                </ul>
              </OuiText>
            </OuiToast>
          </div>
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>
        Toasts should only contain a single action
      </GuideRuleTitle>

      <GuideRule
        description="A toast can have a single action, styled as a standard button.
          If more actions are needed, or if the action is important enough to
          interrupt the user, use a modal instead.">
        <GuideRuleExample
          type="do"
          text="Use only one action per toast and favor a one-word label.
              Align actions to the right, which follows our button guidelines for
              usage within restricted width containers.">
          <OuiToast
            style={{ maxWidth: 300 }}
            color="success"
            title="Your report is complete">
            <div style={{ textAlign: 'right' }}>
              <OuiButton size="s">Download</OuiButton>
            </div>
          </OuiToast>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          text="Use multiple actions. Don't align buttons in toasts to the left.
              This message is better in a confirmation modal.">
          <OuiToast
            style={{ maxWidth: 300 }}
            title="All messages will be deleted"
            color="danger">
            <OuiFlexGroup justifyContent="flexEnd" gutterSize="s">
              <OuiFlexItem grow={false}>
                <OuiButton size="s">Cancel</OuiButton>
              </OuiFlexItem>
              <OuiFlexItem grow={false}>
                <OuiButton size="s" color="danger">
                  Delete
                </OuiButton>
              </OuiFlexItem>
            </OuiFlexGroup>
          </OuiToast>
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>Icons should emphasize actions</GuideRuleTitle>

      <GuideRule description="An icon on the left of the message can help define the message type.">
        <GuideRuleExample
          type="do"
          text="The check icon reinforces that the action succeeded.
                The alert icon helps users understand the message is an error.">
          <div>
            <OuiToast
              style={{ maxWidth: 300 }}
              title="Your dashboard was updated"
              iconType="check"
              color="success"
            />

            <OuiSpacer />

            <OuiToast
              style={{ maxWidth: 300 }}
              title="A dashboard named 'MyDashboard' already exists"
              iconType="alert"
              color="danger"
            />
          </div>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          text="Use icons that are hard to understand. They distract from the message.">
          <OuiToast
            color="primary"
            style={{ maxWidth: 300 }}
            title="Message sent"
            iconType="help"
          />
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>Display one toast at a time</GuideRuleTitle>

      <GuideRule
        description="Users should be able to take
          in all the details from one toast before the next one arrives.">
        <GuideRuleExample type="do" text="Display one toast at a time.">
          <OuiToast
            style={{ maxWidth: 300 }}
            color="primary"
            title="3 new messages"
          />
        </GuideRuleExample>

        <GuideRuleExample type="dont" text="Stack toasts.">
          <div>
            <OuiToast
              style={{ maxWidth: 300 }}
              color="danger"
              title="There was a problem with your node">
              <div style={{ textAlign: 'right' }}>
                <OuiButton size="s">Learn more</OuiButton>
              </div>
            </OuiToast>
            <OuiSpacer />

            <OuiToast
              color="primary"
              style={{ maxWidth: 300 }}
              title="3 new messages"
            />
          </div>
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>Keep messages as short as possible</GuideRuleTitle>

      <GuideRule
        description="For common actions such as create, add, delete, remove, and save,
          include the object type, the object name if available, and the past tense of the action.
          ">
        <GuideRuleExample
          type="do"
          text="Include the object name if it's not too long.
            Use single quotation marks around the object name if it helps clarify meaning.">
          <div>
            <OuiToast
              color="success"
              style={{ maxWidth: 300 }}
              title="User 'Casey Smith' was added"
            />
          </div>
        </GuideRuleExample>

        <GuideRuleExample type="dont" text='Use the generic "Your object."'>
          <OuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="Your object has been saved"
          />
        </GuideRuleExample>
      </GuideRule>

      <GuideRule description="Don't include the word &quot;successfully.&quot; It's implied.">
        <GuideRuleExample
          type="do"
          text="Use this format for a success message.">
          <OuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="Dashboard 'My_dashboard_with_a_very_long_name' was saved"
          />
        </GuideRuleExample>

        <GuideRuleExample type="dont" text='Include "successfully."'>
          <OuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="Dashboard 'My_dashboard' was successfully saved"
          />
        </GuideRuleExample>
      </GuideRule>

      <GuideRule description="For a message about multiple objects, include the object count, but not the names of the objects.">
        <GuideRuleExample type="do" text="Include the object count.">
          <OuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="4 visualizations were deleted"
          />
        </GuideRuleExample>
        <GuideRuleExample
          type="dont"
          text="Overwhelm the user by listing the names of all the objects.">
          <OuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="Visualization 1, Visualization 2, Visualization 3, and Visualization 4 were deleted"
          />
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>
        Use call-to-action buttons when the content needs more room
      </GuideRuleTitle>

      <GuideRule
        description="Occassionally the content of a toast is too involved to fit into the constrained space of a toast.
          This is common in long error messages. In these cases use the toast to deliver the summary of the
          information and use a button to provide a call-to-action for the full message.">
        <GuideRuleExample
          type="do"
          text="Use the toast message to provide a summary and a button to link to the full content">
          <OuiToast
            style={{ maxWidth: 300 }}
            color="danger"
            title="Your visualization has an error">
            <p>The maximum bucket size of 200 was exceeded.</p>
            <div style={{ textAlign: 'right' }}>
              <OuiButton size="s" color="danger" onClick={showModal}>
                See the full error
              </OuiButton>
              {modal}
            </div>
          </OuiToast>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          text="Cram a lot of content into the small space of a toast.">
          <OuiToast
            style={{ maxWidth: 300 }}
            color="danger"
            title="Your visualization has an error">
            <OuiCallOut
              title="The maximum bucket size of 200 was exceeded"
              color="danger">
              <p>An extremely long error trace.</p>
            </OuiCallOut>
          </OuiToast>
        </GuideRuleExample>
      </GuideRule>
    </>
  );
};
