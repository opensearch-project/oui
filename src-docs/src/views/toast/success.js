/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import {
  OuiButton,
  OuiLink,
  OuiToast,
  OuiFlexGroup,
  OuiFlexItem,
} from '../../../../src/components';

export default () => (
  <OuiToast
    title="Created report for dashboard 'Tuba Sales by Region'"
    color="success"
    iconType="check">
    <p>
      While the layout will adjust properly for wrapping titles, they do not
      look particularly good. Similarily, do not use a whole lot of text in your
      body either. At a certain point people will not have enough time to read
      these things. Like, you probably are not even reading this now.
    </p>

    <p>
      And some other stuff on another line, just for kicks. And{' '}
      <OuiLink href="https://oui.opensearch.org/latest/">
        here&rsquo;s a link
      </OuiLink>
      .
    </p>

    <OuiFlexGroup justifyContent="flexEnd" gutterSize="s">
      <OuiFlexItem grow={false}>
        <OuiButton size="s">Download report</OuiButton>
      </OuiFlexItem>
    </OuiFlexGroup>
  </OuiToast>
);
