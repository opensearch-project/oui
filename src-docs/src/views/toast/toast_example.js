/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Link } from 'react-router-dom';

import { renderToHtml } from '../../services';

import { GuideSectionTypes } from '../../components';

import {
  OuiCode,
  OuiToast,
  OuiGlobalToastList,
  OuiGlobalToastListItem,
  OuiText,
} from '../../../../src/components';
import Guidelines from './guidelines';
import toastConfig from './playground';

import ToastList from './toast_list';
const toastListSource = require('./toast_list?raw');
const toastListHtml = renderToHtml(ToastList);
const toastListSnippet = [
  `<OuiGlobalToastList
  toasts={[
    {
      id: 1,
      title: "Example Toast",
      text: <p><!-- Content --></p>,
    }
  ]}
  dismissToast={dismissToast}
  toastLifeTimeMs={6000}/>`,
];

import Default from './default';
const defaultSource = require('./default?raw');
const defaultHtml = renderToHtml(Default);
const defaultToastSnippet = [
  `<OuiToast
  title="Default toast"
  onClose={closeToast}>
  <!-- Raw HTML content -->
  </OuiToast>`,
];

import Info from './info';
const infoSource = require('./info?raw');
const infoHtml = renderToHtml(Info);
const infoToastSnippet = [
  `<OuiToast
  title="Info toast"
  type="info"
  onClose={closeToast}>
  <!-- Raw HTML content -->
  </OuiToast>`,
];

import Success from './success';
const successSource = require('./success?raw');
const successHtml = renderToHtml(Success);
const successToastSnippet = [
  `<OuiToast
  title="Success toast"
  color="success"
  iconType="check"
  onClose={closeToast}>
  <!-- Raw HTML content -->
  </OuiToast>`,
];

import Warning from './warning';
const warningSource = require('./warning?raw');
const warningHtml = renderToHtml(Warning);
const warningToastSnippet = [
  `<OuiToast
  title="Warning toast"
  color="warning"
  iconType="help"
  onClose={closeToast}>
  <!-- Raw HTML content -->
  </OuiToast>`,
];

import Danger from './danger';
const dangerSource = require('./danger?raw');
const dangerHtml = renderToHtml(Danger);
const dangerToastSnippet = [
  `<OuiToast
  title="Danger toast"
  color="danger"
  iconType="alert"
  onClose={closeToast}>
  <!-- Raw HTML content -->
  </OuiToast>`,
];

export const ToastExample = {
  title: 'Toast',
  intro: (
    <OuiText>
      <p>
        Be sure to read the full{' '}
        <Link to="/guidelines/toast">toast usage guidelines</Link>.
      </p>
    </OuiText>
  ),
  sections: [
    {
      title: 'Toast list',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: toastListSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: toastListHtml,
        },
      ],
      props: {
        OuiToast,
        OuiGlobalToastList,
        OuiGlobalToastListItem,
      },
      demo: (
        <>
          <ToastList />
        </>
      ),
      snippet: toastListSnippet,
    },
    {
      title: 'Default',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: defaultSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: defaultHtml,
        },
      ],
      text: (
        <div>
          <p>
            <strong>OuiToast</strong> allows for small notes that appear in the
            bottom right of the screen. They should be used for ephemeral, live
            actions (think <strong>save complete</strong> or{' '}
            <strong>something just finished right now</strong>).
          </p>
          <p>
            They should not be used for historical actions (
            <strong>your report built 30 minutes ago</strong>). This means that
            a user should never be greeted with toasts when starting a session.
            Toasts should be brief and avoid long paragraphs of text or titling.
          </p>
        </div>
      ),
      demo: (
        <div style={{ maxWidth: 320 }}>
          <Default />
        </div>
      ),
      snippet: defaultToastSnippet,
    },
    {
      title: 'Info',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: infoSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: infoHtml,
        },
      ],
      text: (
        <p>
          Setting <OuiCode language="js">type=&quot;info&quot;</OuiCode>.
        </p>
      ),
      demo: (
        <div style={{ maxWidth: 320 }}>
          <Info />
        </div>
      ),
      snippet: infoToastSnippet,
    },
    {
      title: 'Success',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: successSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: successHtml,
        },
      ],
      text: (
        <p>
          Setting <OuiCode language="js">color=&quot;success&quot;</OuiCode>.
        </p>
      ),
      demo: (
        <div style={{ maxWidth: 320 }}>
          <Success />
        </div>
      ),
      snippet: successToastSnippet,
    },
    {
      title: 'Warning',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: warningSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: warningHtml,
        },
      ],
      text: (
        <p>
          Setting <OuiCode language="js">color=&quot;warning&quot;</OuiCode>.
        </p>
      ),
      demo: (
        <div style={{ maxWidth: 320 }}>
          <Warning />
        </div>
      ),
      snippet: warningToastSnippet,
    },
    {
      title: 'Danger',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: dangerSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: dangerHtml,
        },
      ],
      text: (
        <p>
          Setting <OuiCode language="js">color=&quot;danger&quot;</OuiCode>.
        </p>
      ),
      demo: (
        <div style={{ maxWidth: 320 }}>
          <Danger />
        </div>
      ),
      snippet: dangerToastSnippet,
    },
  ],
  guidelines: <Guidelines />,
  playground: toastConfig,
};
