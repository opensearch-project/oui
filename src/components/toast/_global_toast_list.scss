/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/**
 * 1. Allow list to expand as items are added, but cap it at the screen height.
 * 2. Allow some padding for shadow
 */
.ouiGlobalToastList {
  @include ouiScrollBar;

  display: flex;
  flex-direction: column;
  align-items: stretch;
  position: fixed;
  z-index: $ouiZToastList;
  bottom: 0;
  width: $ouiToastWidth + ($ouiSize * 5); /* 2 */
  max-height: 100vh; /* 1 */
  overflow-y: auto;

  // Hide the scrollbar entirely
  scrollbar-width: none;

  &::-webkit-scrollbar {
    width: 0;
    height: 0;
  }

  // The top and bottom padding give height to the list creating a dead-zone effect
  // when there's no toasts in the list, meaning you can't click anything beneath it.
  // Only add the padding if there's content.
  &:not(:empty) {
    padding: $ouiSize;
  }
}

.ouiGlobalToastList--right:not(:empty) {
  right: 0;
  padding-left: $ouiSize * 4; /* 2 */
}

.ouiGlobalToastList--left:not(:empty) {
  left: 0;
  padding-right: $ouiSize * 4; /* 2 */
}

.ouiGlobalToastListItem {
  margin-bottom: $ouiSize;
  animation: $ouiAnimSpeedNormal ouiShowToast $ouiAnimSlightResistance;
  opacity: 1;

  /**
   * 1. justify-content: flex-end interferes with overflowing content, so we'll use this to push
   *    items to the bottom instead.
   */
  &:first-child {
    margin-top: auto; /* 1 */
  }

  &:last-child {
    margin-bottom: 0;
  }

  &.ouiGlobalToastListItem-isDismissed {
    transition: opacity $ouiAnimSpeedNormal;
    opacity: 0;
  }
}

@keyframes ouiShowToast {
  from {
    transform: translateY($ouiSizeL) scale(.9);
    opacity: 0;
  }

  to {
    transform: translateY(0) scale(1);
    opacity: 1;
  }
}

@include ouiBreakpoint('xs', 's') {
  /**
   * 1. Mobile we make these 100%. Matching change happens on the item as well.
   */
  .ouiGlobalToastList:not(:empty) {
    left: 0;
    padding-left: $ouiSize;
    padding-right: $ouiSize;
    width: 100%; /* 1 */
  }
}
