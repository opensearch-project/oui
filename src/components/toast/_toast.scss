/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

.ouiToast {
  border: $ouiBorderThin;

  // This overwrites some of the border above
  @include ouiBottomShadowLarge($adjustBorders: true);

  position: relative;
  padding: $ouiSize;
  background-color: $ouiColorEmptyShade;
  width: 100%;

  &:hover .ouiToast__closeButton,
  &:focus .ouiToast__closeButton {
    opacity: 1;
  }
}

/**
  * 1. Fit button to icon.
  */
.ouiToast__closeButton {
  position: absolute;
  top: $ouiSize;
  right: $ouiSize;
  line-height: 0; /* 1 */
  appearance: none;
  opacity: 0;
  transition: opacity $ouiAnimSpeedFast $ouiAnimSlightResistance;

  svg {
    fill: tintOrShade($ouiTitleColor, 50%, 70%);
  }

  &:hover {
    svg {
      fill: $ouiTitleColor;
    }
  }

  &:focus {
    background-color: $ouiFocusBackgroundColor;
    opacity: 1;

    svg {
      fill: $ouiColorPrimary;
    }
  }
}

// Modifier naming and colors.
$ouiToastTypes: (
  primary: makeGraphicContrastColor($ouiColorPrimary, $ouiColorEmptyShade),
  success: makeGraphicContrastColor($ouiColorSecondary, $ouiColorEmptyShade),
  warning: makeGraphicContrastColor($ouiColorWarning, $ouiColorEmptyShade),
  danger: makeGraphicContrastColor($ouiColorDanger, $ouiColorEmptyShade),
);

// Create button modifiers based upon the map.
@each $name, $color in $ouiToastTypes {
  .ouiToast--#{$name} {
    border-top: 2px solid $color;
  }
}

/**
 * 1. Align icon with first line of title text if it wraps.
 * 2. Apply margin to all but last item in the flex.
 * 3. Account for close button.
 */
.ouiToastHeader {
  padding-right: $ouiSizeL; /* 3 */
  display: flex;
  align-items: baseline; /* 1 */

  > * + * {
    margin-left: $ouiSizeS; /* 2 */
  }
}

/**
  * 1. Vertically center icon with first line of title.
  */
.ouiToastHeader__icon {
  flex: 0 0 auto;
  fill: $ouiTitleColor;
  transform: translateY(2px); /* 1 */
}

.ouiToastHeader__title {
  @include ouiTitle('xs');
  @include ouiTextBreakWord;

  font-weight: $ouiFontWeightLight;
}

.ouiToastHeader--withBody {
  margin-bottom: $ouiSizeS;
}

/**
 * 1. Prevent long lines from overflowing.
 */
.ouiToastBody {
  word-wrap: break-word; /* 1 */
}


/* OUI -> EUI Aliases */
$euiToastTypes: $ouiToastTypes;

/* End of Aliases */
