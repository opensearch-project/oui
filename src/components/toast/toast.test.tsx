/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { requiredProps } from '../../test';

import { COLORS, OuiToast } from './toast';

describe('OuiToast', () => {
  test('is rendered', () => {
    const { container } = render(
      <OuiToast {...requiredProps} title="test title">
        <p>Hi</p>
      </OuiToast>
    );

    expect(container).toMatchSnapshot();
  });

  describe('Props', () => {
    describe('title', () => {
      test('is rendered', () => {
        const { container } = render(<OuiToast title="toast title" />);
        expect(container).toMatchSnapshot();
      });
    });

    describe('color', () => {
      COLORS.forEach((color) => {
        test(`${color} is rendered`, () => {
          const { container } = render(
            <OuiToast color={color} title="test title" />
          );
          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('iconType', () => {
      test('is rendered', () => {
        const { container } = render(
          <OuiToast iconType="user" title="test title" />
        );
        expect(container).toMatchSnapshot();
      });
    });

    describe('onClose', () => {
      test('is called when the close button is clicked', () => {
        const onCloseHandler = jest.fn();

        render(<OuiToast onClose={onCloseHandler} title="test title" />);

        act(() => {
          fireEvent.click(screen.getByTestId('toastCloseButton'));
        });

        expect(onCloseHandler).toHaveBeenCalledTimes(1);
      });
    });
  });
});
