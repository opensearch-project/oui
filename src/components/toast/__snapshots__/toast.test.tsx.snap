// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiToast Props color danger is rendered 1`] = `
<div>
  <div
    class="ouiToast ouiToast--danger"
  >
    <p
      class="ouiScreenReaderOnly"
    >
      A new notification appears
    </p>
    <div
      aria-label="Notification"
      class="ouiToastHeader"
      data-test-subj="ouiToastHeader"
    >
      <span
        class="ouiToastHeader__title"
      >
        test title
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiToast Props color primary is rendered 1`] = `
<div>
  <div
    class="ouiToast ouiToast--primary"
  >
    <p
      class="ouiScreenReaderOnly"
    >
      A new notification appears
    </p>
    <div
      aria-label="Notification"
      class="ouiToastHeader"
      data-test-subj="ouiToastHeader"
    >
      <span
        class="ouiToastHeader__title"
      >
        test title
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiToast Props color success is rendered 1`] = `
<div>
  <div
    class="ouiToast ouiToast--success"
  >
    <p
      class="ouiScreenReaderOnly"
    >
      A new notification appears
    </p>
    <div
      aria-label="Notification"
      class="ouiToastHeader"
      data-test-subj="ouiToastHeader"
    >
      <span
        class="ouiToastHeader__title"
      >
        test title
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiToast Props color warning is rendered 1`] = `
<div>
  <div
    class="ouiToast ouiToast--warning"
  >
    <p
      class="ouiScreenReaderOnly"
    >
      A new notification appears
    </p>
    <div
      aria-label="Notification"
      class="ouiToastHeader"
      data-test-subj="ouiToastHeader"
    >
      <span
        class="ouiToastHeader__title"
      >
        test title
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiToast Props iconType is rendered 1`] = `
<div>
  <div
    class="ouiToast"
  >
    <p
      class="ouiScreenReaderOnly"
    >
      A new notification appears
    </p>
    <div
      aria-label="Notification"
      class="ouiToastHeader"
      data-test-subj="ouiToastHeader"
    >
      <span
        aria-hidden="true"
        class="ouiToastHeader__icon"
        data-ouiicon-type="user"
      />
      <span
        class="ouiToastHeader__title"
      >
        test title
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiToast Props title is rendered 1`] = `
<div>
  <div
    class="ouiToast"
  >
    <p
      class="ouiScreenReaderOnly"
    >
      A new notification appears
    </p>
    <div
      aria-label="Notification"
      class="ouiToastHeader"
      data-test-subj="ouiToastHeader"
    >
      <span
        class="ouiToastHeader__title"
      >
        toast title
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiToast is rendered 1`] = `
<div>
  <div
    aria-label="aria-label"
    class="ouiToast testClass1 testClass2"
    data-test-subj="test subject string"
  >
    <p
      class="ouiScreenReaderOnly"
    >
      A new notification appears
    </p>
    <div
      aria-label="Notification"
      class="ouiToastHeader ouiToastHeader--withBody"
      data-test-subj="ouiToastHeader"
    >
      <span
        class="ouiToastHeader__title"
      >
        test title
      </span>
    </div>
    <div
      class="ouiText ouiText--small ouiToastBody"
    >
      <p>
        Hi
      </p>
    </div>
  </div>
</div>
`;
